File: services/balance-sheet/engine/ratios.py
"""
Financial ratio computation.
Computes key balance-sheet ratios from normalized extraction periods.
"""

from __future__ import annotations

from models.schemas import BalanceSheetRatios, ExtractionPeriod


def compute_ratios(periods: list[ExtractionPeriod]) -> BalanceSheetRatios:
    """
    Compute financial ratios from the most recent extraction period.
    Returns a BalanceSheetRatios object.
    """
    if not periods:
        return BalanceSheetRatios()

    # Use the most recent period
    latest = periods[0]
    items = latest.line_items

    current_ratio: float | None = None
    debt_to_equity: float | None = None
    working_capital: float | None = None
    goodwill_to_assets: float | None = None
    intangibles_to_assets: float | None = None

    # Current ratio = current assets / current liabilities
    if items.total_current_assets is not None and items.total_current_liabilities is not None:
        if items.total_current_liabilities != 0:
            current_ratio = items.total_current_assets / items.total_current_liabilities

    # Debt to equity = total liabilities / total equity
    if items.total_liabilities is not None and items.total_equity is not None:
        if items.total_equity != 0:
            debt_to_equity = items.total_liabilities / items.total_equity

    # Working capital = current assets - current liabilities
    if items.total_current_assets is not None and items.total_current_liabilities is not None:
        working_capital = items.total_current_assets - items.total_current_liabilities

    # Goodwill to total assets
    if items.goodwill is not None and items.total_assets is not None:
        if items.total_assets != 0:
            goodwill_to_assets = items.goodwill / items.total_assets

    # Intangibles to total assets
    if items.intangible_assets is not None and items.total_assets is not None:
        if items.total_assets != 0:
            intangibles_to_assets = items.intangible_assets / items.total_assets

    return BalanceSheetRatios(
        current_ratio=_round_safe(current_ratio, 3),
        debt_to_equity=_round_safe(debt_to_equity, 3),
        working_capital=_round_safe(working_capital, 2),
        goodwill_to_assets=_round_safe(goodwill_to_assets, 4),
        intangibles_to_assets=_round_safe(intangibles_to_assets, 4),
    )


def _round_safe(value: float | None, decimals: int) -> float | None:
    """Round a value safely, returning None if input is None."""
    if value is None:
        return None
    return round(value, decimals)

File: src/lib/balance-sheet/ratios.ts
// Financial ratio computation — all derived strictly from balance sheet line items.

import { NormalizedPeriod, ComputedRatios } from './types';

function roundSafe(v: number | null, d: number): number | null {
  return v == null ? null : Math.round(v * 10 ** d) / 10 ** d;
}

function safeDiv(n: number | null | undefined, d: number | null | undefined): number | null {
  if (n == null || d == null || d === 0) return null;
  return n / d;
}

export function computeRatios(periods: NormalizedPeriod[]): ComputedRatios {
  if (!periods.length) {
    return {
      currentRatio: null, quickRatio: null, cashRatio: null,
      debtToEquity: null, debtToAssets: null,
      workingCapital: null, workingCapitalToAssets: null,
      goodwillToAssets: null, intangiblesToAssets: null, tangibleEquity: null,
      arShareOfCurrentAssets: null, inventoryShareOfCurrentAssets: null,
      apShareOfCurrentLiab: null, shortTermDebtShareOfDebt: null,
    };
  }

  const i = periods[0].lineItems;

  const currentAssets = i.total_current_assets;
  const currentLiab = i.total_current_liabilities;
  const cash = i.cash_and_equivalents;
  const sti = i.short_term_investments;
  const ar = i.accounts_receivable;
  const inv = i.inventory;
  const ap = i.accounts_payable;
  const std = i.short_term_debt;
  const ltd = i.long_term_debt;
  const totalAssets = i.total_assets;
  const totalLiab = i.total_liabilities;
  const equity = i.total_equity;
  const goodwill = i.goodwill ?? 0;
  const intangibles = i.intangible_assets ?? 0;

  const currentRatio = safeDiv(currentAssets, currentLiab);

  // Quick ratio = (CA - inventory) / CL.  Fallback to cash+sti+ar when CA/inventory missing.
  let quickNum: number | null = null;
  if (currentAssets != null && inv != null) quickNum = currentAssets - inv;
  else if (cash != null || sti != null || ar != null) {
    quickNum = (cash ?? 0) + (sti ?? 0) + (ar ?? 0);
  }
  const quickRatio = safeDiv(quickNum, currentLiab);

  const cashRatio = safeDiv((cash ?? 0) + (sti ?? 0), currentLiab);

  const debtToEquity = safeDiv(totalLiab, equity);
  const debtToAssets = safeDiv(totalLiab, totalAssets);

  const workingCapital =
    currentAssets != null && currentLiab != null ? currentAssets - currentLiab : null;
  const workingCapitalToAssets = safeDiv(workingCapital, totalAssets);

  const goodwillToAssets = safeDiv(goodwill, totalAssets);
  const intangiblesToAssets = safeDiv(goodwill + intangibles, totalAssets);

  const tangibleEquity =
    equity != null ? equity - goodwill - intangibles : null;

  const arShareOfCurrentAssets = safeDiv(ar, currentAssets);
  const inventoryShareOfCurrentAssets = safeDiv(inv, currentAssets);
  const apShareOfCurrentLiab = safeDiv(ap, currentLiab);

  const totalDebt = (std ?? 0) + (ltd ?? 0);
  const shortTermDebtShareOfDebt =
    totalDebt > 0 && std != null ? std / totalDebt : null;

  return {
    currentRatio: roundSafe(currentRatio, 3),
    quickRatio: roundSafe(quickRatio, 3),
    cashRatio: roundSafe(cashRatio, 3),
    debtToEquity: roundSafe(debtToEquity, 3),
    debtToAssets: roundSafe(debtToAssets, 3),
    workingCapital: roundSafe(workingCapital, 2),
    workingCapitalToAssets: roundSafe(workingCapitalToAssets, 4),
    goodwillToAssets: roundSafe(goodwillToAssets, 4),
    intangiblesToAssets: roundSafe(intangiblesToAssets, 4),
    tangibleEquity: roundSafe(tangibleEquity, 2),
    arShareOfCurrentAssets: roundSafe(arShareOfCurrentAssets, 4),
    inventoryShareOfCurrentAssets: roundSafe(inventoryShareOfCurrentAssets, 4),
    apShareOfCurrentLiab: roundSafe(apShareOfCurrentLiab, 4),
    shortTermDebtShareOfDebt: roundSafe(shortTermDebtShareOfDebt, 4),
  };
}
